internals/plan/plan.go

```diff
@@ -42,11 +68,68 @@ const (
 	defaultCheckThreshold = 3
 )
 
+// layerExtensions keeps a map of registered extensions.
+var layerExtensions = map[string]LayerSectionExtension{}
+
+// layerBuiltins represents all the built-in layer sections. This list is used
+// for identifying built-in fields in this package. It is unit tested to match
+// the YAML fields exposed in the Layer type, to catch inconsistencies.
+var layerBuiltins = []string{"summary", "description", "services", "checks", "log-targets"}
+
+// RegisterExtension adds a plan schema extension. All registrations must be
+// done before the plan library is used.
+func RegisterExtension(field string, ext LayerSectionExtension) {
+	if slices.Contains(layerBuiltins, field) {
+		panic(fmt.Sprintf("internal error: extension %q already used as built-in field", field))
+	}
+	if _, ok := layerExtensions[field]; ok {
+		panic(fmt.Sprintf("internal error: extension %q already registered", field))
+	}
+	layerExtensions[field] = ext
+}
+
+// UnregisterExtension removes a plan schema extension. This is only
+// intended for use by tests during cleanup.
+func UnregisterExtension(field string) {
+	delete(layerExtensions, field)
+}
+
 type Plan struct {
 	Layers     []*Layer              `yaml:"-"`
 	Services   map[string]*Service   `yaml:"services,omitempty"`
 	Checks     map[string]*Check     `yaml:"checks,omitempty"`
 	LogTargets map[string]*LogTarget `yaml:"log-targets,omitempty"`
+
+	Sections map[string]LayerSection `yaml:",inline"`
```

Comment:
This seems a bit awkward to read. I realize it's mimicking what was there before, with Services, Checks, etc, but now it seems even more weird as we have Layers, Sections, and individual Sections in the same level. This is okay if it's temporary, but is it?  And if so, we need a comment in here properly spelling out the plan so the next person can make sense of the current state. Can we do better so it's more sensible meanwhile?

Reply:
I previously played with having:

```
type Plan struct {
	Layers     []*Layer              `yaml:"-"`
	combined *Layer               `yaml:",inline"
}
```

The main issue I observed was that you lose the ability to simply access content from the plan directly, unless you add helper methods, and this propagated far into the codebase:

```
func (p *Plan) Services() map[string]*Service {
    return p.combined.Services
}
```

If we do that, then perhaps waiting until we migrated the 3 built-in sections out, so that it looks like this:

```
type Plan struct {
	Layers     []*Layer              `yaml:"-"`
	Sections map[string]LayerSection `yaml:",inline"`
}
```

@benhoyt Perhaps we can brainstorm what could be nice here?